File: SortColorsLeetcode75.java
//Time Complexity: O(n)
//Space Complexity: O(1)
//Leetcode: yes.

public class SortColorsLeetcode75 {
        public void sortColors(int[] nums) {
            if (nums == null) {
                return;
            }

            int fp = 0;
            int sp = nums.length -1;
            int index = 0;
            /* If element is 0 then swap it at first place and move both pointers
             * If element is 1 do nothing
             * If element is 2 then swao it with the last element and move the second pointer
             * do not move the index as the new element can be 2 again so we need to check that again
             */
            while (index <= sp) {
                if (nums[index] == 0) {
                    swap(nums, index, fp);
                    fp++;
                    index++;
                } else if (nums[index] == 2) {
                    swap(nums, index, sp);
                    sp--;
                } else {
                    index++;
                }
            }
            return;
        }

        public void swap(int[] nums, int i, int j) {
            int temp = nums[i];
            nums[i] = nums[j];
            nums[j] = temp;
        }
}

File: Sort_colors_leetcode75.py
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # One pass TC - O(n) SC - O(1)
        n = len(nums)
        c = 0
        l = 0
        r = n - 1
        while c <= r:
            if nums[c] == 0:
                nums[c], nums[l] = nums[l], nums[c]
                c += 1
                l += 1

            elif nums[c] == 2:
                nums[c], nums[r] = nums[r], nums[c]
                r -= 1

            else:  # if c == 1 then just move c pointer
                c += 1

        '''
        # Easy - Using 2 pass, counting R, W, B and overwriting nums
        # TC - O(n) SC - O(1)
        R = 0
        W = 0
        B = 0
        n = len(nums)
        for i in range(len(nums)):
            if nums[i] == 0:
                R += 1
            elif nums[i] == 1:
                W += 1
            elif nums[i] == 2:
                B += 1

        i = 0
        while R > 0 and i < n:
            R -= 1
            nums[i] = 0
            i += 1
        while W > 0 and i < n:
            W -= 1
            nums[i] = 1
            i += 1
        while B > 0 and i < n:
            B -= 1
            nums[i] = 2
            i += 1
        '''

